core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -325,24 +325,26 @@ final case class NonEmptyList[+A](head: A, tail: List[A]) {
     *
     * {{{
     * scala> import cats.data.NonEmptyList
+    * scala> import cats.instances.boolean._
     * scala> val nel = NonEmptyList.of(12, -2, 3, -5)
     * scala> nel.groupBy(_ >= 0)
     * res0: Map[Boolean, cats.data.NonEmptyList[Int]] = Map(false -> NonEmptyList(-2, -5), true -> NonEmptyList(12, 3))
     * }}}
     */
-  def groupBy[B](f: A => B): Map[B, NonEmptyList[A]] = {
-    val m = mutable.Map.empty[B, mutable.Builder[A, List[A]]]
+  def groupBy[B](f: A => B)(implicit B: Order[B]): Map[B, NonEmptyList[A]] = {
+    var m = TreeMap.empty[B, mutable.Builder[A, List[A]]](B.toOrdering)
+
     for { elem <- toList } {
-      m.getOrElseUpdate(f(elem), List.newBuilder[A]) += elem
-    }
-    val b = immutable.Map.newBuilder[B, NonEmptyList[A]]
-    for { (k, v) <- m } {
-      val head :: tail = v.result // we only create non empty list inside of the map `m`
-      b += ((k, NonEmptyList(head, tail)))
+      val k = f(elem)
+
+      m.get(k) match {
+        case None => m += ((k, List.newBuilder[A] += elem))
+        case Some(builder) => builder += elem
+      }
     }
-    b.result
-  }
 
+    m.map { case (k, v) => (k, NonEmptyList.fromListUnsafe(v.result)) }
```

Comment:
can we ascribe a type to this to make sure we are not converting to using hashCode and equals at the end?
```scala
 m.map { case (k, v) => (k, NonEmptyList.fromListUnsafe(v.result)) } : TreeMap[B, NonEmptyList[A]]
```

Reply:
Sure. One further thought. Shall we also change the return type to `SortedMap`? That would explain why we require an `Order` instance for `B`.